src/BlockNLPModels.jl

```diff
@@ -0,0 +1,213 @@
+      return new(0,0,0,0)
+  end    
+end
+
+"""
+    AbstractBlockModel{T, S} <: AbstractNLPModel{T, S}    
+A data type to store block subproblems of the form
+```math
+\\begin{aligned}
+  \\min_{x \\in \\mathbb{R^{m_i}}} \\quad & f_i(x_i) \\\\
+  \\mathrm{subject \\, to} \\quad & c_{ij} (x_i) \\leq 0 \\quad \\forall j \\in \\mathcal{C}_i, \\\\
+\\end{aligned}
+where ``i`` is the index of the block.
+"""
+mutable struct AbstractBlockModel{T, S} <: AbstractNLPModel{T, S}
+  meta::NLPModelMeta{T, S}
+  counters::Counters
+  problem_block::AbstractNLPModel
+  idx::Int
+  var_idx::UnitRange{Int} # Indices of variables wrt the full space model
+  con_idx::UnitRange{Int} # Indices of constraints wrt the full space model
+  linking_constraint_id::Vector{Int} # ID of linking constraints that connect this block with other blocks 
+end
+
+
+"""
+    LinearLinkConstraint <: AbstractLinearLinkConstraint
+A data type to store linear link constraints of the form
+```math
+\\begin{aligned}
+  \\sum\\limits_{i \\in \\mathcal{B}} A_i x_i = b
+\\end{aligned}
+"""
+mutable struct LinearLinkConstraint <: AbstractLinearLinkConstraint
+  linking_blocks::Vector{SparseMatrixCSC{Float64,Int}}
+  idx::UnitRange{Int}
+  link_map::Dict{Int, Vector{Int}} # constraint => blocks, i.e., which constraint connects which blocks
+  RHS_vector::Vector{Float64}
+end
+
+mutable struct NLLinkConstraint <: AbstractNLLinkConstraint
+end
+
+"""
+    BlockNLPModel <: AbstractBlockNLPModel
+A data type designed to store nonlinear optimization models of the form
+```math
+\\begin{aligned}
+  \\min_{x \\in \\mathbb{R^{m_i}}} \\quad & \\sum\\limits_{i \\in \\mathcal{B}} f_i(x_i) \\\\
+  \\mathrm{subject \\, to} \\quad & c_{ij} (x_i) \\leq 0 \\quad \\forall i \\in \\mathcal{B}, j \\in \\mathcal{C}_i \\\\
+  & \\sum\\limits_{i \\in \\mathcal{B}} A_i x_i = b,
+\\end{aligned}
+where ``\\mathcal{B}`` is the set of variable blocks.
+
+---
+
+    BlockNLPModel()
+
+Initializes an empty `BlockNLPModel`. 
+"""
+mutable struct BlockNLPModel <: AbstractBlockNLPModel
+  problem_size::BlockNLP_Counters
+  blocks::Vector{AbstractBlockModel}
+  linking_constraints::Vector{Union{AbstractLinearLinkConstraint, AbstractNLLinkConstraint}}
+  function BlockNLPModel()
+    return new(
+      BlockNLP_Counters(),
+      Vector{AbstractBlockModel}(),
+      Vector{Union{AbstractLinearLinkConstraint, AbstractNLLinkConstraint}}()
+    )
+  end
+end
+
+"""
+  add_block(
+    block_nlp::AbstractBlockNLPModel, 
+    nlp::AbstractNLPModel
+  )
+
+Adds a block subproblem to a `BlockNLPModel`.
+The `BlockNLPModel` needs to be initialized before this function can be called.
+
+# Arguments
+
+- `block_nlp::AbstractBlockNLPModel`: name of the BlockNLPModel to which a subproblem is to be added.
+- `nlp::AbstractNLPModel`: the subproblem.
+"""
+function add_block(block_nlp::AbstractBlockNLPModel, nlp::AbstractNLPModel)
+  block_nlp.problem_size.block_counter += 1
+
+  temp = deepcopy(block_nlp.problem_size.var_counter)+1
```

Comment:
Do we really need this machinery?

Reply:
I am using this to assign a unique index to all block variables. Having a unique indes assigned to every variable was very useful while writing ``fullspace.jl``, ``dualized_block.jl``. I am not sure if there is a smarter way that can avoid this. I am open to suggestions.